### PR TITLE
Support word split

### PR DESCRIPTION
**Pull request by singlecoder**
### Please check if the PR fulfills these requirements

- [ ] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### word-split

yuque_diagram (3)

### CODE REVIEW

File: packages/core/src/2d/text/TextUtils.ts [hunk 1]
```diff
@@ -110,52 +110,135 @@ export class TextUtils {
...
       for (let j = 0, m = subText.length; j < m; ++j) {
         const char = subText[j];
         const charInfo = TextUtils._getCharInfo(char, fontString, subFont);
+        const charCode = char.charCodeAt(0);
+        // 32 is space.
+        const isNotEnglish = charCode === 32 || charCode > 255;
         const { w, offsetY } = charInfo;
         const halfH = charInfo.h * 0.5;
         const ascent = halfH + offsetY;
         const descent = halfH - offsetY;
```
Comment (cptbtptpbcptdtptp) [inline packages/core/src/2d/text/TextUtils.ts:128]: Why not use baseline based `ascent` and `descent`?

File: packages/core/src/2d/text/TextUtils.ts [hunk 2]
```diff
@@ -110,52 +110,135 @@ export class TextUtils {
...
-            lineMaxSizes.push({
-              ascent,
-              descent,
-              size: ascent + descent
-            });
+
+        if (isNotEnglish) {
+          // If it is an English word before, need to handle the previous English word and chars.
+          if (curInEnglishWord) {
+            if (charsWidth + wordCharsWidth > wrapWidth) {
```
Comment (cptbtptpbcptdtptp) [inline packages/core/src/2d/text/TextUtils.ts:133]: I think the definition of `wrapWidth` should not be `renderer.width`, he should use a new variable definition, like `wordWrapWidth`.

**Comment by cptbtptpbcptdtptp** [comment]
- A space cannot be the first character of a line
- Cannot be displayed when `wordWrapWidth` is 0

File: packages/core/src/2d/text/TextUtils.ts (hunk 2 above)
Comment (singlecoder) [reply packages/core/src/2d/text/TextUtils.ts:133]: I think this will make things more complex, but the local bounds should discuss.

**Comment by singlecoder** [comment]
> 



> 
When enableWrapping is true, width is 0, display nothing is expected.

File: packages/core/src/2d/text/TextUtils.ts [hunk 3]
```diff
@@ -110,52 +110,154 @@ export class TextUtils {
...
+          }
+
+          // Handle cur char.
+          if (charsWidth + w > wrapWidth) {
+            if (charsWidth === 0) {
+              this._pushCharsToLines(lines, lineWidths, lineMaxSizes, char, w, ascent, descent);
+              isNotFirstLine = true;
+            } else {
+              this._pushCharsToLines(lines, lineWidths, lineMaxSizes, chars, charsWidth, maxAscent, maxDescent);
+              isNotFirstLine = true;
```
Comment (cptbtptpbcptdtptp) [inline packages/core/src/2d/text/TextUtils.ts:167]: Can be combined with the above.

**Comment by cptbtptpbcptdtptp** [comment]
image
Arranged according to your logic：
```
  static measureTextWithWrap(renderer: TextRenderer): TextMetrics {
    const { fontSize, fontStyle, _subFont: subFont } = renderer;
    const { name } = renderer.font;
    const fontString = TextUtils.getNativeFontString(name, fontSize, fontStyle);
    const fontSizeInfo = TextUtils.measureFont(fontString);
    const subTexts = renderer.text.split(/(?:\r\n|\r|\n)/);
    const lines = new Array<string>();
    const lineWidths = new Array<number>();
    const lineMaxSizes = new Array<FontSizeInfo>();
    const { _pixelsPerUnit } = Engine;
    const lineHeight = fontSizeInfo.size + renderer.lineSpacing * _pixelsPerUnit;
    const wrapWidth = renderer.width * _pixelsPerUnit;
    let width = 0;
    for (let i = 0, n = subTexts.length; i < n; ++i) {
      const subText = subTexts[i];
      let lineChars = "";
      let lineWidth = 0;
      let lineAscent = 0;
      let lineDescent = 0;
      let wordsChars = "";
      let wordsWidth = 0;
      let wordsAscent = 0;
      let wordsDescent = 0;
      let isNotFirstLine = false;
      for (let j = 0, m = subText.length; j < m; j++) {
        const char = subText[j];
        const bSpace = char === " ";
        if (bSpace && lineWidth === 0 && wordsWidth === 0 && isNotFirstLine) {
          continue;
        }
        const charInfo = TextUtils._getCharInfo(char, fontString, subFont);
        const { w: charWidth, offsetY } = charInfo;
        const halfH = charInfo.h >> 1;
        const ascent = halfH + offsetY;
        const descent = halfH - offsetY;
        const canFormWords = !bSpace && char.charCodeAt(0) <= 255;
        if (canFormWords) {
          if (wordsWidth + charWidth > wrapWidth) {
            if (lineWidth > 0) {
              this._pushCharsToLines(lines, lineWidths, lineMaxSizes, lineChars, lineWidth, lineAscent, lineDescent);
              lineChars = "";
              lineWidth = lineAscent = lineDescent = 0;
            }
            this._pushCharsToLines(lines, lineWidths, lineMaxSizes, wordsChars, wordsWidth, wordsAscent, wordsDescent);
            isNotFirstLine = true;
            wordsChars = char;
            wordsWidth = charWidth;
            wordsAscent = ascent;
            wordsDescent = descent;
          } else {
            wordsChars += char;
            wordsWidth += charWidth;
            wordsAscent = Math.max(wordsAscent, ascent);
            wordsDescent = Math.max(wordsDescent, descent);
          }
        } else {
          if (wordsWidth > 0) {
            if (wordsWidth + lineWidth > wrapWidth) {
              this._pushCharsToLines(lines, lineWidths, lineMaxSizes, lineChars, lineWidth, lineAscent, lineDescent);
              isNotFirstLine = true;
              lineChars = wordsChars;
              lineWidth = wordsWidth;
              lineAscent = wordsAscent;
              lineDescent = wordsDescent;
            } else {
              lineChars += wordsChars;
              lineWidth += wordsWidth;
              lineAscent = Math.max(lineAscent, wordsAscent);
              lineDescent = Math.max(lineDescent, wordsDescent);
            }
            wordsChars = "";
            wordsWidth = wordsAscent = wordsDescent = 0;
          }
          if (lineWidth + charWidth > wrapWidth) {
            this._pushCharsToLines(lines, lineWidths, lineMaxSizes, lineChars, lineWidth, lineAscent, lineDescent);
            isNotFirstLine = true;
            lineChars = "";
            lineWidth = 0;
            lineAscent = 0;
            lineDescent = 0;
          }
          if (!bSpace || lineWidth !== 0 || !isNotFirstLine) {
            lineChars += char;
            lineWidth += charWidth;
            lineAscent = Math.max(ascent, lineAscent);
            lineDescent = Math.max(descent, lineDescent);
          }
        }
      }
      if (wordsWidth > 0) {
        if (wordsWidth + lineWidth > wrapWidth) {
          this._pushCharsToLines(lines, lineWidths, lineMaxSizes, lineChars, lineWidth, lineAscent, lineDescent);
          lineChars = wordsChars;
          lineWidth = wordsWidth;
          lineAscent = wordsAscent;
          lineDescent = wordsDescent;
        } else {
          lineChars += wordsChars;
          lineWidth += wordsWidth;
          lineAscent = Math.max(lineAscent, wordsAscent);
          lineDescent = Math.max(lineDescent, wordsDescent);
        }
        wordsChars = "";
        wordsWidth = wordsAscent = wordsDescent = 0;
      }
      if (lineWidth > 0) {
        this._pushCharsToLines(lines, lineWidths, lineMaxSizes, lineChars, lineWidth, lineAscent, lineDescent);
        lineChars = "";
        lineWidth = lineAscent = lineDescent = 0;
      }
    }

    let height = renderer.height * _pixelsPerUnit;
    if (renderer.overflowMode === OverflowMode.Overflow) {
      height = lineHeight * lines.length;
    }

    return {
      width,
      height,
      lines,
      lineWidths,
      lineHeight,
      lineMaxSizes
    };
  }
```

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]
Do some extreme cases to verify the rigor of the logic (at least you can verify whether it crashes)

File: packages/core/src/2d/text/TextUtils.ts [hunk 4]
```diff
@@ -110,54 +110,135 @@ export class TextUtils {
...
       if (charsWidth > 0) {
-        lines.push(chars);
-        lineWidths.push(charsWidth);
-        lineMaxSizes.push({
-          ascent: maxAscent,
-          descent: maxDescent,
-          size: maxAscent + maxDescent
-        });
+        this._pushCharsToLines(lines, lineWidths, lineMaxSizes, chars, charsWidth, maxAscent, maxDescent);
+        isNotFirstLine = true;
```
Comment (cptbtptpbcptdtptp) [inline packages/core/src/2d/text/TextUtils.ts:240]: Can delete.

File: packages/core/src/2d/text/TextUtils.ts [hunk 5]
```diff
@@ -110,54 +110,135 @@ export class TextUtils {
...
+          this._pushCharsToLines(
+            lines,
+            lineWidths,
+            lineMaxSizes,
+            wordChars,
+            wordCharsWidth,
+            wordMaxAscent,
+            wordMaxDescent
+          );
+          isNotFirstLine = true;
```
Comment (cptbtptpbcptdtptp) [inline packages/core/src/2d/text/TextUtils.ts:227]: Can delete.

**Review by cptbtptpbcptdtptp — APPROVED** [review]
+1

File: packages/core/src/2d/text/TextUtils.ts [hunk 6]
```diff
@@ -110,54 +110,133 @@ export class TextUtils {
...
         const char = subText[j];
         const charInfo = TextUtils._getCharInfo(char, fontString, subFont);
+        const charCode = char.charCodeAt(0);
+        // 32 is space.
+        const isSpace = charCode === 32;
+        if (isSpace && isNotFirstLine && chars.length === 0 && wordChars.length === 0) {
+          continue;
+        }
+
+        const isNotJoinWords = isSpace || charCode > 255;
```
[the remark this answers is not in the document]
Comment (GuoLei1990) [reply packages/core/src/2d/text/TextUtils.ts:131]: image

**Comment by GuoLei1990** [comment]
The current text alignment is centered, I think left alignment and top alignment are more commonly used

**Pull request merged**